### PR TITLE
Dyninst: add conflicts for unsupported compilers

### LINKED ISSUE #17874: Installation issue: dyninst on NERSC Cori
Using `spack@develop` (3-Aug, 4c97a0ea1cbcb42fff4cb2183552b6b6c441d791) on NERSC Cori (`cray-cnl7-haswell`) with `%intel@19.1.1.217`:

The Spack environment used to install `dyninst` is attached below along with the full build log.

```
$> spack -e . install
...

...
==> Installing dyninst
==> dyninst: Executing phase: 'cmake'
==> dyninst: Executing phase: 'build'
...
    920     [ 38%] Linking CXX shared library libpcontrol.so
     921     cd /tmp/eugeneswalker/spack-stage/spack-stage-dyninst-10.1.0-dvejlhtkjzq4g4434747ta6g7kgelop6/spack-build/proccontrol && /global/u1/l/eugeneswalker/spack/spack_path_placeholder/
             spack_path_placeholder/spack_path_placeholder/spack_path_placeholder/spack_path_placeholder/spack_path_placeholder/spack_path_placeholder/spack_path_placeholder/spack_
             path_placeholder/spack_path_placeholder/spack_path_placeholder/spack_path_placeholder/spack_path_placeholder/spack_path_placeholder/spack_path_placeholder/spack_path_p
             laceholder/spack_path_placeholder/spack_path_placeholder/spack_path_placeholder/spack_path_placeholder/spack_path_placeholder/spack_path_placeholder/spack_/cray-cnl7-h
             aswell/intel-19.1.1.217/cmake-3.16.5-xlzhc4yr46hivtofz6axdfyokymjzb6x/bin/cmake -E cmake_link_script CMakeFiles/pcontrol.dir/link.txt --verbose=1
     922     /global/u1/l/eugeneswalker/spack/lib/spack/env/intel/icpc -fPIC  -std=c++11  -O2 -g   -Wl,--no-undefined  -shared -Wl,-soname,libpcontrol.so.10.1 -o libpcontrol.so.10.1.0 C
             MakeFiles/pcontrol.dir/src/event.C.o CMakeFiles/pcontrol.dir/src/generator.C.o CMakeFiles/pcontrol.dir/src/handler.C.o CMakeFiles/pcontrol.dir/src/mailbox.C.o CMakeFil
             es/pcontrol.dir/src/process.C.o CMakeFiles/pcontrol.dir/src/pcerrors.C.o CMakeFiles/pcontrol.dir/src/procpool.C.o CMakeFiles/pcontrol.dir/src/irpc.C.o CMakeFiles/pcont
             rol.dir/src/response.C.o CMakeFiles/pcontrol.dir/src/resp.C.o CMakeFiles/pcontrol.dir/src/memcache.C.o CMakeFiles/pcontrol.dir/src/procset.C.o CMakeFiles/pcontrol.dir/
             src/processplat.C.o CMakeFiles/pcontrol.dir/src/loadLibrary/injector.C.o CMakeFiles/pcontrol.dir/src/loadLibrary/codegen.C.o CMakeFiles/pcontrol.dir/src/loadLibrary/co
             degen-x86.C.o CMakeFiles/pcontrol.dir/src/loadLibrary/codegen-ppc.C.o CMakeFiles/pcontrol.dir/src/loadLibrary/codegen-aarch64.C.o CMakeFiles/pcontrol.dir/src/x86_proce
             ss.C.o CMakeFiles/pcontrol.dir/src/ppc_process.C.o CMakeFiles/pcontrol.dir/src/arm_process.C.o CMakeFiles/pcontrol.dir/src/pcmachsyscall.C.o CMakeFiles/pcontrol.dir/sr
             c/sysv.C.o CMakeFiles/pcontrol.dir/src/int_thread_db.C.o CMakeFiles/pcontrol.dir/src/mmapalloc.C.o CMakeFiles/pcontrol.dir/src/linux.C.o CMakeFiles/pcontrol.dir/src/un
             ix.C.o CMakeFiles/pcontrol.dir/src/notify_pipe.C.o CMakeFiles/pcontrol.dir/__/common/src/dthread-unix.C.o CMakeFiles/pcontrol.dir/__/common/src/dthread.C.o CMakeFiles/
             pcontrol.dir/src/loadLibrary/codegen-linux.C.o   -L/global/u1/l/eugeneswalker/spack/spack_path_placeholder/spack_path_placeholder/spack_path_placeholder/spack_path_placehol
             der/spack_path_placeholder/spack_path_placeholder/spack_path_placeholder/spack_path_placeholder/spack_path_placeholder/spack_path_placeholder/spack_path_placeholder/sp
             ack_path_placeholder/spack_path_placeholder/spack_path_placeholder/spack_path_placeholder/spack_path_placeholder/spack_path_placeholder/spack_path_placeholder/spack_pa
             th_placeholder/spack_path_placeholder/spack_path_placeholder/spack_path_placeholder/spack_/cray-cnl7-haswell/intel-19.1.1.217/boost-1.73.0-suzkrf3mnjn2fuilhvb3avdybwry
             jms3/lib  -L/global/u1/l/eugeneswalker/spack/spack_path_placeholder/spack_path_placeholder/spack_path_placeholder/spack_path_placeholder/spack_path_placeholder/spack_path_p
             laceholder/spack_path_placeholder/spack_path_placeholder/spack_path_placeholder/spack_path_placeholder/spack_path_placeholder/spack_path_placeholder/spack_path_placeho
             lder/spack_path_placeholder/spack_path_placeholder/spack_path_placeholder/spack_path_placeholder/spack_path_placeholder/spack_path_placeholder/spack_path_placeholder/s
             pack_path_placeholder/spack_path_placeholder/spack_/cray-cnl7-haswell/intel-19.1.1.217/intel-tbb-2020.3-pb7wn67upluau7pmao2y6voologmzjlt/lib  -L/global/u1/l/eugeneswalker/s
             pack/spack_path_placeholder/spack_path_placeholder/spack_path_placeholder/spack_path_placeholder/spack_path_placeholder/spack_path_placeholder/spack_path_placeholder/s
             pack_path_placeholder/spack_path_placeholder/spack_path_placeholder/spack_path_placeholder/spack_path_placeholder/spack_path_placeholder/spack_path_placeholder/spack_p
             ath_placeholder/spack_path_placeholder/spack_path_placeholder/spack_path_placeholder/spack_path_placeholder/spack_path_placeholder/spack_path_placeholder/spack_path_pl
             aceholder/spack_/cray-cnl7-haswell/intel-19.1.1.217/elfutils-0.180-hubfank2wm3mtw36czn5e7enbjpyhnqq/lib  -Wl,-rpath,/global/u1/l/eugeneswalker/spack/spack_path_placeholder/
             spack_path_placeholder/spack_path_placeholder/spack_path_placeholder/spack_path_placeholder/spack_path_placeholder/spack_path_placeholder/spack_path_placeholder/spack_
             path_placeholder/spack_path_placeholder/spack_path_placeholder/spack_path_placeholder/spack_path_placeholder/spack_path_placeholder/spack_path_placeholder/spack_path_p
             laceholder/spack_path_placeholder/spack_path_placeholder/spack_path_placeholder/spack_path_placeholder/spack_path_placeholder/spack_path_placeholder/spack_/cray-cnl7-h
             aswell/intel-19.1.1.217/boost-1.73.0-suzkrf3mnjn2fuilhvb3avdybwryjms3/lib:/global/u1/l/eugeneswalker/spack/spack_path_placeholder/spack_path_placeholder/spack_path_placehol
             der/spack_path_placeholder/spack_path_placeholder/spack_path_placeholder/spack_path_placeholder/spack_path_placeholder/spack_path_placeholder/spack_path_placeholder/sp
             ack_path_placeholder/spack_path_placeholder/spack_path_placeholder/spack_path_placeholder/spack_path_placeholder/spack_path_placeholder/spack_path_placeholder/spack_pa
             th_placeholder/spack_path_placeholder/spack_path_placeholder/spack_path_placeholder/spack_path_placeholder/spack_/cray-cnl7-haswell/intel-19.1.1.217/intel-tbb-2020.3-p
             b7wn67upluau7pmao2y6voologmzjlt/lib:/global/u1/l/eugeneswalker/spack/spack_path_placeholder/spack_path_placeholder/spack_path_placeholder/spack_path_placeholder/spack_path_
             placeholder/spack_path_placeholder/spack_path_placeholder/spack_path_placeholder/spack_path_placeholder/spack_path_placeholder/spack_path_placeholder/spack_path_placeh
             older/spack_path_placeholder/spack_path_placeholder/spack_path_placeholder/spack_path_placeholder/spack_path_placeholder/spack_path_placeholder/spack_path_placeholder/
             spack_path_placeholder/spack_path_placeholder/spack_path_placeholder/spack_/cray-cnl7-haswell/intel-19.1.1.217/elfutils-0.180-hubfank2wm3mtw36czn5e7enbjpyhnqq/lib:/tmp
             /eugeneswalker/spack-stage/spack-stage-dyninst-10.1.0-dvejlhtkjzq4g4434747ta6g7kgelop6/spack-build/symtabAPI:/tmp/eugeneswalker/spack-stage/spack-stage-dyninst-10.1.0-dvejlhtkjz
             q4g4434747ta6g7kgelop6/spack-build/dwarf:/tmp/eugeneswalker/spack-stage/spack-stage-dyninst-10.1.0-dvejlhtkjzq4g4434747ta6g7kgelop6/spack-build/common:/tmp/eugeneswalker/spack-s
             tage/spack-stage-dyninst-10.1.0-dvejlhtkjzq4g4434747ta6g7kgelop6/spack-build/elf:::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::
             :::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::: ../symtabAPI/libsymtabAPI.so.10.1.0 -ldl /gl
             obal/u1/l/eugeneswalker/spack/spack_path_placeholder/spack_path_placeholder/spack_path_placeholder/spack_path_placeholder/spack_path_placeholder/spack_path_placeholder/spac
             k_path_placeholder/spack_path_placeholder/spack_path_placeholder/spack_path_placeholder/spack_path_placeholder/spack_path_placeholder/spack_path_placeholder/spack_path
             _placeholder/spack_path_placeholder/spack_path_placeholder/spack_path_placeholder/spack_path_placeholder/spack_path_placeholder/spack_path_placeholder/spack_path_place
             holder/spack_path_placeholder/spack_/cray-cnl7-haswell/intel-19.1.1.217/boost-1.73.0-suzkrf3mnjn2fuilhvb3avdybwryjms3/lib/libboost_atomic-mt.so /global/u1/l/eugeneswalker/s
             pack/spack_path_placeholder/spack_path_placeholder/spack_path_placeholder/spack_path_placeholder/spack_path_placeholder/spack_path_placeholder/spack_path_placeholder/s
             pack_path_placeholder/spack_path_placeholder/spack_path_placeholder/spack_path_placeholder/spack_path_placeholder/spack_path_placeholder/spack_path_placeholder/spack_p
             ath_placeholder/spack_path_placeholder/spack_path_placeholder/spack_path_placeholder/spack_path_placeholder/spack_path_placeholder/spack_path_placeholder/spack_path_pl
             aceholder/spack_/cray-cnl7-haswell/intel-19.1.1.217/boost-1.73.0-suzkrf3mnjn2fuilhvb3avdybwryjms3/lib/libboost_chrono-mt.so /global/u1/l/eugeneswalker/spack/spack_path_plac
             eholder/spack_path_placeholder/spack_path_placeholder/spack_path_placeholder/spack_path_placeholder/spack_path_placeholder/spack_path_placeholder/spack_path_placeholde
             r/spack_path_placeholder/spack_path_placeholder/spack_path_placeholder/spack_path_placeholder/spack_path_placeholder/spack_path_placeholder/spack_path_placeholder/spac
             k_path_placeholder/spack_path_placeholder/spack_path_placeholder/spack_path_placeholder/spack_path_placeholder/spack_path_placeholder/spack_path_placeholder/spack_/cra
             y-cnl7-haswell/intel-19.1.1.217/boost-1.73.0-suzkrf3mnjn2fuilhvb3avdybwryjms3/lib/libboost_date_time-mt.so /global/u1/l/eugeneswalker/spack/spack_path_placeholder/spack_pat
             h_placeholder/spack_path_placeholder/spack_path_placeholder/spack_path_placeholder/spack_path_placeholder/spack_path_placeholder/spack_path_placeholder/spack_path_plac
             eholder/spack_path_placeholder/spack_path_placeholder/spack_path_placeholder/spack_path_placeholder/spack_path_placeholder/spack_path_placeholder/spack_path_placeholde
             r/spack_path_placeholder/spack_path_placeholder/spack_path_placeholder/spack_path_placeholder/spack_path_placeholder/spack_path_placeholder/spack_/cray-cnl7-haswell/in
             tel-19.1.1.217/boost-1.73.0-suzkrf3mnjn2fuilhvb3avdybwryjms3/lib/libboost_filesystem-mt.so /global/u1/l/eugeneswalker/spack/spack_path_placeholder/spack_path_placeholder/sp
             ack_path_placeholder/spack_path_placeholder/spack_path_placeholder/spack_path_placeholder/spack_path_placeholder/spack_path_placeholder/spack_path_placeholder/spack_pa
             th_placeholder/spack_path_placeholder/spack_path_placeholder/spack_path_placeholder/spack_path_placeholder/spack_path_placeholder/spack_path_placeholder/spack_path_pla
             ceholder/spack_path_placeholder/spack_path_placeholder/spack_path_placeholder/spack_path_placeholder/spack_path_placeholder/spack_/cray-cnl7-haswell/intel-19.1.1.217/b
             oost-1.73.0-suzkrf3mnjn2fuilhvb3avdybwryjms3/lib/libboost_system-mt.so /global/u1/l/eugeneswalker/spack/spack_path_placeholder/spack_path_placeholder/spack_path_placeholder
             /spack_path_placeholder/spack_path_placeholder/spack_path_placeholder/spack_path_placeholder/spack_path_placeholder/spack_path_placeholder/spack_path_placeholder/spack
             _path_placeholder/spack_path_placeholder/spack_path_placeholder/spack_path_placeholder/spack_path_placeholder/spack_path_placeholder/spack_path_placeholder/spack_path_
             placeholder/spack_path_placeholder/spack_path_placeholder/spack_path_placeholder/spack_path_placeholder/spack_/cray-cnl7-haswell/intel-19.1.1.217/boost-1.73.0-suzkrf3m
             njn2fuilhvb3avdybwryjms3/lib/libboost_thread-mt.so /global/u1/l/eugeneswalker/spack/spack_path_placeholder/spack_path_placeholder/spack_path_placeholder/spack_path_placehol
             der/spack_path_placeholder/spack_path_placeholder/spack_path_placeholder/spack_path_placeholder/spack_path_placeholder/spack_path_placeholder/spack_path_placeholder/sp
             ack_path_placeholder/spack_path_placeholder/spack_path_placeholder/spack_path_placeholder/spack_path_placeholder/spack_path_placeholder/spack_path_placeholder/spack_pa
             th_placeholder/spack_path_placeholder/spack_path_placeholder/spack_path_placeholder/spack_/cray-cnl7-haswell/intel-19.1.1.217/boost-1.73.0-suzkrf3mnjn2fuilhvb3avdybwry
             jms3/lib/libboost_timer-mt.so -lpthread ../common/libcommon.so.10.1.0 -ltbbmalloc_proxy -ltbbmalloc -ltbb -Wl,-rpath-link,/tmp/eugeneswalker/spack-stage/spack-stage-dyninst
             -10.1.0-dvejlhtkjzq4g4434747ta6g7kgelop6/spack-build/dwarf:/tmp/eugeneswalker/spack-stage/spack-stage-dyninst-10.1.0-dvejlhtkjzq4g4434747ta6g7kgelop6/spack-build/elf
     923     icpc: command line warning #10121: overriding '-xCORE-AVX2' with '-march=haswell'
     924     icpc: command line warning #10121: overriding '-xCORE-AVX2' with '-march=haswell'
     925
  >> 926     ": internal error: ** The compiler has encountered an unexpected problem.
     927     ** Segmentation violation signal raised. **
     928     Access violation or stack overflow. Please contact Intel Support for assistance.
     929
  >> 930     icpc: error #10106: Fatal error in /global/common/cori_cle7/software/intel/parallel_studio_xe_2020_update1_cluster_edition/compilers_and_libraries_2020.1.217/linux/bin
             /intel64/mcpcom, terminated by segmentation violation
  >> 931     icpc: error #10014: problem during multi-file optimization compilation (code 1)
  >> 932     make[2]: *** [proccontrol/CMakeFiles/pcontrol.dir/build.make:549: proccontrol/libpcontrol.so.10.1.0] Error 1
     933     make[2]: Leaving directory '/tmp/eugeneswalker/spack-stage/spack-stage-dyninst-10.1.0-dvejlhtkjzq4g4434747ta6g7kgelop6/spack-build'
  >> 934     make[1]: *** [CMakeFiles/Makefile2:1153: proccontrol/CMakeFiles/pcontrol.dir/all] Error 2
     935     make[1]: *** Waiting for unfinished jobs....
     936     [ 38%] Building CXX object parseAPI/CMakeFiles/parseAPI.dir/src/CFGModifier.C.o
```

Full build log: [dyninst-build-out.txt](https://github.com/spack/spack/files/5022978/dyninst-build-out.txt)
Spack environment: [spack.yaml.txt](https://github.com/spack/spack/files/5022979/spack.yaml.txt)

@lee218llnl @jgalarowicz @frankwillmore @scottwittenburg @shahzebsiddiqui @mwkrentel @mpbelhorn

### PR DESCRIPTION
**Pull request by hainest**
Closes #18219
Closes #17874

### CODE REVIEW

File: var/spack/repos/builtin/packages/dyninst/package.py [hunk 1]
```diff
@@ -61,6 +61,16 @@ class Dyninst(CMakePackage):
     patch('v9.3.2-auto.patch', when='@9.3.2 %gcc@:4.7.99')
     patch('tribool.patch', when='@9.3.0:10.0.0 ^boost@1.69:')
 
+    # We currently only build with gcc
+    conflicts('%clang')
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/dyninst/package.py:68]: There is also a `%apple-clang` compiler now

Comment (hainest) [reply var/spack/repos/builtin/packages/dyninst/package.py:68]: Is that case not handled by #18219?

Comment (adamjstewart) [reply var/spack/repos/builtin/packages/dyninst/package.py:68]: Technically it is since you can't build `%apple-clang` on a non-macOS system.

**Review by adamjstewart — APPROVED** [review]

File: var/spack/repos/builtin/packages/dyninst/package.py (hunk 1 above)
Comment (hainest) [reply var/spack/repos/builtin/packages/dyninst/package.py:68]: Ok. Maybe we should merge that one first. Then I'll rebase this, if necessary.

Comment (adamjstewart) [reply var/spack/repos/builtin/packages/dyninst/package.py:68]: I can't merge mine until someone reviews and approves it

Comment (adamjstewart) [reply var/spack/repos/builtin/packages/dyninst/package.py:68]: Honestly, you could just add that conflict to your PR and I can close my PR.

Comment (hainest) [reply var/spack/repos/builtin/packages/dyninst/package.py:68]: Done.

**Review by alalazo — APPROVED** [review]

**Pull request merged**